auditable/auditable.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package auditable
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"go/types"
+	"strings"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+var (
+	apiPackagePath = "github.com/mattermost/mattermost-server/v6/api4"
+)
+
+var Analyzer = &analysis.Analyzer{
+	Name: "auditable",
+	Doc:  "check if auditable interface is satisfied when passing structs to audit records",
+	Run:  run,
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	if pass.Pkg.Path() != apiPackagePath {
+		return nil, nil
+	}
+
+	for _, file := range pass.Files {
+		commentMap := make(map[token.Pos]*ast.CommentGroup)
+		for i := range file.Comments {
+			commentMap[file.Comments[i].Pos()] = file.Comments[i]
+		}
+
+		ast.Inspect(file, func(n ast.Node) bool {
+			expr, ok := n.(*ast.CallExpr)
+			if !ok {
+				return true
+			}
+
+			if expr.Fun == nil {
+				return true
+			}
+
+			fn, ok := expr.Fun.(*ast.SelectorExpr)
+			if !ok {
+				return true
+			}
+
+			if fn.Sel != nil && fn.Sel.Name != "AddEventParameter" {
+				return true
+			}
```

Comment:
This is nice and I wonder if at some point we could extract some of this logic and make it re-usable since I could see a similar pattern applied again in the future. Essentially we could generalize a bunch of boilerplate code and only pass a test function (`isAuditable` in this case) and some parameters to make this applicable to future cases as well.

Reply:
Added logic for reusing.